### DISCUSSION TITLE
Any way to export a MTLCommandBuffer?

### DISCUSSION THREAD

**Discussion by llGuy** [KhronosGroup/MoltenVK#1924] [Q&A]
Hi there,

This may be a silly question, but is there any way to access the underlying MTLCommandBuffer?

I want to use Metal Performance Shaders in my Vulkan Application and that API requires access to an MTLCommandBuffer - the closest thing I see to this is MTLCommandQueue.

Looking at the source code, I don't think this would be possible - seems like commands don't get encoded immediately, but all at once at the end of recording the command buffer?

Regards,
Luc

**Comment by llGuy** [comment]
Follow up: If this truly isn't possible, I want to see whether this is the correct approach for achieving what I want.

What I want:
I have a bunch of Vulkan resources (buffer containing matrices, etc...) and a command buffer in which I recorded some render passes and compute passes. I want to insert a metal performance shader matrix multiplication in here. The only way to do this, is through encoding into a MTLCommandQueue.

Possible approach:
Could I create a VkSemaphore, export the internal MTLSharedEvent object and encode a signalEvent (with what I believe should be signaledValue+1 - seems like that just gets incremented after each commit?) after the MPS shader, and then freely use that semaphore to force the next command buffers (which contain the render passes and such) to wait?

I'm just wondering if this would break anything.

Thanks

**Reply by billhollings** [reply]
Unlike `VkCommandBuffer`, a `MTLCommandBuffer` is a transient object, which in general, would be impractical and dangerous to export. So, as you surmised, exporting a `MTLCommandQueue` and then creating a `MTLCommandBuffer` from it is your best bet.

As far as synchronization goes, what you're attempting sounds reasonable (although unexplored so far).
